tests/components/modbus/conftest.py

```diff
@@ -37,63 +32,101 @@ def __init__(self, register_words):
-    if device is None:
-        pytest.fail("CONFIG failed, see output")
-    return entity_id, now, device
-
-
-async def run_base_read_test(
-    entity_id,
-    hass,
-    use_mock_hub,
-    register_type,
+    array_name_discovery,
+    array_name_old_config,
     register_words,
     expected,
-    now,
+    method_discovery=False,
+    check_config_only=False,
+    config_modbus=None,
+    scan_interval=None,
 ):
-    """Run test for given config."""
-    # Setup inputs for the sensor
-    read_result = ReadResult(register_words)
-    if register_type == CALL_TYPE_COIL:
-        use_mock_hub.read_coils.return_value = read_result
-    elif register_type == CALL_TYPE_DISCRETE:
-        use_mock_hub.read_discrete_inputs.return_value = read_result
-    elif register_type == CALL_TYPE_REGISTER_INPUT:
-        use_mock_hub.read_input_registers.return_value = read_result
-    else:  # CALL_TYPE_REGISTER_HOLDING
-        use_mock_hub.read_holding_registers.return_value = read_result
-
-    # Trigger update call with time_changed event
-    with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
-        async_fire_time_changed(hass, now)
-        await hass.async_block_till_done()
-
-    # Check state
-    state = hass.states.get(entity_id).state
-    assert state == expected
+    """Run test on device for given config."""
+
+    if config_modbus is None:
+        config_modbus = {
+            DOMAIN: {
+                CONF_NAME: DEFAULT_HUB,
+                CONF_TYPE: "tcp",
+                CONF_HOST: "modbusTest",
+                CONF_PORT: 5001,
+            },
+        }
+
+    mock_sync = mock.MagicMock()
+    with mock.patch(
+        "homeassistant.components.modbus.modbus.ModbusTcpClient", return_value=mock_sync
+    ), mock.patch(
+        "homeassistant.components.modbus.modbus.ModbusSerialClient",
+        return_value=mock_sync,
+    ), mock.patch(
+        "homeassistant.components.modbus.modbus.ModbusUdpClient", return_value=mock_sync
+    ):
+
+        # Setup inputs for the sensor
+        read_result = ReadResult(register_words)
+        mock_sync.read_coils.return_value = read_result
+        mock_sync.read_discrete_inputs.return_value = read_result
+        mock_sync.read_input_registers.return_value = read_result
+        mock_sync.read_holding_registers.return_value = read_result
+
+        # mock timer and add modbus platform with devices (new config)
+        now = dt_util.utcnow()
+        with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
+            if method_discovery:
+                # setup modbus which in turn does setup for the devices
+                if config_device is not None:
+                    config_modbus[DOMAIN].update(
+                        {array_name_discovery: [{**config_device}]}
+                    )
+                await async_load_platform(
+                    hass, DOMAIN, DOMAIN, config_modbus[DOMAIN], config_modbus
```

Comment:
There's no modbus platform. The platforms are cover, sensor, switch etc in the modbus integration.

For the platforms that have been migrated to be loaded via discovery, like cover, it's enough to set up the modbus integration with `async_setup_component`.

Reply:
you got me there :-) First I understood the difference between "integration" and "platform"; but then I save the calls "setup_component" which seems to indicate something in between.

Anyhow I have corrected it, and thanks for explaining the difference.

Reply:
We used to call integrations "components".